mongo/jobs.go

```diff
@@ -25,104 +25,81 @@ func (m *JobStore) UnlockJob(ctx context.Context, lockID string) {
-				"reindex_started_time": jobStartTime,
-			}
-			log.Info(ctx, "found an existing reindex job in progress", logData)
-			return ErrExistingJobInProgress
+	// get job with state "in-progress" and the most recent "reindex_started" time
+	err := s.DB(m.Database).C(m.JobsCollection).Find(bson.M{"state": "in-progress"}).Sort("-reindex_started").One(&job)
+	if err != nil {
+		if err == mgo.ErrNotFound {
+			log.Info(ctx, "no reindex jobs with state in progress")
+			return nil
 		}
+		log.Error(ctx, "failed to get job with state in-progress and the most recent reindex_started time", err)
+		return err
 	}
 
-	defer func() {
-		if err := iter.Close(); err != nil {
-			log.Error(ctx, "error closing iterator", err)
+	// checkFromTime is the time of configured variable "MaxReindexJobRuntime" from now
+	checkFromTime := time.Now().Add(-1 * m.cfg.MaxReindexJobRuntime)
+
+	// check if start time of the job in progress is later than the checkFromTime but earlier than now
+	if job.ReindexStarted.After(checkFromTime) && job.ReindexStarted.Before(time.Now()) {
+		logData := log.Data{
+			"id":              job.ID,
+			"state":           job.State,
+			"reindex_started": job.ReindexStarted,
 		}
-	}()
+		log.Info(ctx, "found an existing reindex job in progress", logData)
+		return ErrExistingJobInProgress
+	}
 
 	return nil
 }
 
 // CreateJob creates a new job, with the given search index name, in the collection, and assigns default values to its attributes
-func (m *JobStore) CreateJob(ctx context.Context, searchIndexName string) (*models.Job, error) {
-	logData := log.Data{"search_index_name": searchIndexName}
+func (m *JobStore) CreateJob(ctx context.Context, job models.Job) error {
+	logData := log.Data{"job": job}
 	log.Info(ctx, "creating reindex job in mongo DB", logData)
 
 	s := m.Session.Copy()
 	defer s.Close()
 
-	// create a new job
-	newJob, err := models.NewJob(ctx, searchIndexName)
+	// insert job into mongoDB
+	err := s.DB(m.Database).C(m.JobsCollection).Insert(job)
 	if err != nil {
-		log.Error(ctx, "failed to create new job", err, logData)
-		return nil, err
-	}
-
-	// checks if there exists a reindex job with the same id as the new job
-	err = m.validateJobIDUnique(ctx, newJob.ID)
-	if err != nil {
-		logData["jobID"] = newJob.ID
-		log.Error(ctx, "failed to validate job id is unique", err, logData)
-		return nil, err
-	}
-
-	// insert new job into mongoDB
-	err = s.DB(m.Database).C(m.JobsCollection).Insert(*newJob)
-	if err != nil {
-		logData["new_job"] = newJob
-		log.Error(ctx, "failed to insert new job into mongo DB", err, logData)
-		return nil, err
+		log.Error(ctx, "failed to insert job into mongo DB", err, logData)
+		return err
 	}
 
-	return newJob, err
+	return nil
 }
 
-// validateJobIDUnique checks if there exists a reindex job in mongo with the given id
-func (m *JobStore) validateJobIDUnique(ctx context.Context, id string) error {
+func (m *JobStore) findJob(ctx context.Context, jobID string) (models.Job, error) {
```

Comment:
Use pointer for `models.Job`

Reply:
And update job variable below (line 88) and if there is an error then we can return nil instead of creating an object that is empty an unused by the handler

Reply:
I have addressed this in https://github.com/ONSdigital/dp-search-reindex-api/pull/105 already